test/test_protocol.py

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2018-present, Facebook, Inc.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+#
+import dns
+import dns.message
+import struct
+import unittest
+
+from unittest.mock import patch
+from dohproxy.server_protocol import DNSClientProtocolTCP
+
+
+class TCPTestCase(unittest.TestCase):
+
+    def setUp(self):
+        """
+        prepare the dns response
+        """
+        self.dnsq = dns.message.make_query(
+            'www.example.com',
+            dns.rdatatype.ANY)
+        self.dnsr = dns.message.make_response(self.dnsq)
+        self.response = self.dnsr.to_wire()
+
+    def r_receive_helper(self, dnsr):
+        """
+        fut attribute as a list to store the answer from receive helper
+        """
+        self.fut.append(dnsr)
+
+    def r_eof_received(self):
+        """
+        here we raise a general expection when we discard broken message
+        """
+        if len(self.buffer) > 0:
+            raise Exception
+
+    @patch.object(DNSClientProtocolTCP, 'receive_helper', new=r_receive_helper)
+    @patch.object(DNSClientProtocolTCP, 'eof_received', new=r_eof_received)
+    def test_single_valid(self):
```

Comment:
just use mocks and check that they are called or not with given content.

this should look like:
```python
    @patch.object(DNSClientProtocolTCP, 'receive_helper')
    def test_single_valid(self, m_rcv):
        data = struct.pack("!H", len(self.response)) + self.response
        self.client_tcp = DNSClientProtocolTCP(self.dnsq, [], '10.0.0.0')
        self.client_tcp.data_received(data)
        m_rcv.assert_called_with(self.client_tcp, self.response)
```
or something along this line